src/NHibernate/Impl/ExpressionQueryImpl.cs

```diff
@@ -74,11 +74,15 @@ protected override IQueryExpression ExpandParameters(IDictionary<string, TypedVa
 			}
 
 			//TODO: Do we need to translate expression one more time here?
+			// This is not much an issue anymore: ExpressionQueryImpl are currently created only with NhLinqExpression
+			// which do cache their translation. By the way, the false transmitted below as filter parameter to
+			// Translate is wrong, since this ExpandParameters may also be called from ExpressionFilterImpl.
+			// But that does not matter for NhLinqExpression.
 			var newTree = ParameterExpander.Expand(QueryExpression.Translate(Session.Factory, false), map);
```

Comment:
ParameterExpander will duplicate the node again.

Reply:
It needs to duplicate each parameter node potentially many times, due to the possibility of having many aliases for the same parameter. So it would awkward to change it for not duplicating by example for the last current parameter's alias while duplicating for previous ones.

Then the other option would be to avoid the duplication in `Translate` specifically for this case. I am not sure it is worth it. (After all the benchmark still show a gain despite this unneeded duplication for this specific case.) But I may still add a commit for looking how it looks to implement this optimization.

Reply:
Well it will not look very good.

We need to either add a new interface method useful only for one of the three current implementations and meaningless for the others (and incurring the bloat of using at first an extension method for avoiding a breaking change), or to cast to `NhLinqExpression` and use a specific method on it.

Additional commit done. All to be squashed together, or with second and/or third commit dropped if considered better not to have them.